refactor: Adjust pipeline orchestration in `PipelineExecutor`

robinholzi opened the pull request:
# Motivation

The current state of `pipeline_executor.py` doesn't follow clear sequential or logical order. Also, the setup doesn't allow to automatically profile and log pipeline stage execution. Restructuring the orchestration of a pipeline in a PipelineExecutor here.

# Changes

- Split all logical pipeline stages into separate `PipelineExecutor` member functions and merge unused PipelineStages (like `RUN_TRAINING` and `TRAINING_DONE`). This allows to have a 1:1 mapping between logical pipeline stages and member functions.
- Sort the pipeline stage functions sequentially according to their position in the pipeline
- Add `@pipeline_stage(...)` decorator which wraps every pipeline stage function. When calling the decorated function we record meta information for `logging` (post-experiment analysis) and `tracking` (for online feedback) purposes. This includes `start` & `end` timestamps as well as the current "dataset/sample time". This is especially useful in experiments to be able to sort/plot different measurements by the time of the processed samples. 
- Move all state (that is not enclosed in controller objects like `grpc`) into a central `ExecutionState` dataclass.
- Adjusted tests accordingly, removing code duplication in tests by parameterizing tests with `pytest.mark.parameterize(...)`
- Some renaming and other minor adjustments

robinholzi commented:
Resolving the merge conflicts was an odyssey here🫨
@XianzheMa @MaxiBoether Let me know in case I dropped any of the recent changes on `main` (I mostly reimplemented your changes in `(test_)pipeline_executor.py`

XianzheMa commented on modyn/supervisor/internal/grpc/enums.py:
```diff
@@ -18,27 +18,68 @@ class PipelineStatus(StrEnum):
+    MAIN = "main"
+
+    REPLAY_DATA = "replay_data"
+    SERVE_ONLINE = "wait_for_new_data"
```
I thought we could have a strictly identical variable name to the literal name?

XianzheMa commented on modyn/supervisor/internal/grpc_handler.py:
```diff
@@ -276,69 +278,47 @@ def start_training(
             optimizer_config["param_groups"] = []
             for param_group in optimizer["param_groups"]:
-                config_dict = param_group["config"] if "config" in param_group else {}
+                config_dict = param_group.get("config") or {}
```
`get` has the second parameter which is returned if key is absent

robinholzi replied:
I know, the two variants are not equivalent though and unfortunately, the key absence isn't enough here:

`{"key": None}.get("key", {}) == None`
`{"key": None}.get("key") or {} == {}`

XianzheMa commented on modyn/supervisor/internal/grpc_handler.py:
```diff
@@ -276,69 +278,47 @@ def start_training(
 
         lr_scheduler_configs = {}
-        if "lr_scheduler" in pipeline_config["training"]:
+        if pipeline_config["training"].get("lr_scheduler"):
```
why this change?

robinholzi replied:
Using the `in` operator for dicts isn't safe in python as it semantically doesn't always reflect what is intended. Dumping pydantic models can e.g. yield `None` values for dict fields. `in` treats the keys as present while no value is present.

XianzheMa commented on modyn/supervisor/internal/grpc_handler.py:
```diff
@@ -276,69 +278,47 @@ def start_training(
             lr_scheduler_configs = pipeline_config["training"]["lr_scheduler"]
-            if "config" not in lr_scheduler_configs:
-                lr_scheduler_configs["config"] = {}
+            lr_scheduler_configs["config"] = lr_scheduler_configs.get("config", None)
```
these are not equal. Why the change?

XianzheMa commented on modyn/supervisor/internal/grpc_handler.py:
```diff
@@ -276,69 +278,47 @@ def start_training(
-            epochs_per_trigger = pipeline_config["training"]["epochs_per_trigger"]
-        else:
-            epochs_per_trigger = 1
+        epochs_per_trigger = pipeline_config["training"].get("epochs_per_trigger") or 1
```
above can both be re-written with the second arg of `get`

robinholzi replied:
No

XianzheMa commented on modyn/supervisor/internal/grpc_handler.py:
```diff
@@ -276,69 +278,47 @@ def start_training(
+        seed = pipeline_config["training"].get("seed", None)
+        tokenizer = pipeline_config["data"].get("tokenizer", None)
+        transform_list = pipeline_config["data"].get("transformations") or []
+        label_transformer = pipeline_config["data"].get("label_transformer_function") or ""
```
also question on why the second arg of `get` is not used

XianzheMa commented on modyn/supervisor/internal/grpc_handler.py:
```diff
@@ -654,11 +599,6 @@ def wait_for_evaluation_completion(
-            pipeline_stage_msg(
-                PipelineStage.EVALUATION_COMPLETED, MsgType.ID, id_submsg(IdType.TRAINING, training_id), True
-            )
-        )
```
I think most of changes in this file are unrelated with the purpose of this PR. And `if key not in dict` does not look worse than `if dict.get(key)`. To make sure we didn't accidentally introduce something wrong in the request preparation, I suggest we keep the old code.

XianzheMa replied:
In general I advocate the principle "if there is not a really justified reason for a change, then keep it the same".

A pure code style change just adds more work to the reviewer and more risk to uncaught bugs (for example [here](https://github.com/eth-easl/modyn/pull/435/files#r1606098572), hard to catch by the reviewer), because our unit tests certainly do not cover a lot of cases. (Just passing all the tests does not mean it is risk-free). And the old code has less risk to go wrong because we have used it for long.

I hope my thoughts make sense!

XianzheMa replied:
Ah now I see your point in https://github.com/eth-easl/modyn/pull/435/files#r1603030383. Yes in that way this makes sense. Sorry for the misunderstanding!

XianzheMa reviewed (changes requested):
Thanks a lot for the efforts! Yes with the refactored code it looks cleaner, but it's better not to refactor everything in one PR for the reasons:

1. It is difficult to ensure "with naked eyes" the functionality equivalence before and after the changes. I am worried about introducing unforeseen bugs (also missing tests).
2. `pipeline_executor.py` is one of the most important files as it orchestrates the entire lifecycle of a pipeline. We very much rely on its correctness to run experiments. The current version has been used for a long time and thus has a high credibility on correctness. Ensuring its credibility on correctness to run experiments (after all this is our main purpose), should be prioritized over code style refactoring, unless a really needed feature has to be introduced.

In general I really like how it looks after this PR, but I am afraid we miss something during this big refactoring, and then experiments go wrong unnoticed (and we need to spend more time to find it out). Can we split it into multiple PRs that each changes a part of it, so that we have a smooth refactoring?

robinholzi commented on modyn/supervisor/internal/grpc/enums.py:
```diff
@@ -18,27 +18,68 @@ class PipelineStatus(StrEnum):
+    MAIN = "main"
+
+    REPLAY_DATA = "replay_data"
+    SERVE_ONLINE = "wait_for_new_data"
```
I renamed the enum with my IDE from another place and forgot about the string value! Thanks for the hint!

MaxiBoether commented:
hey Robin, thanks for the PR! I will go through it _probably_ on Thursday/Friday. I have to prepare teaching and a NeurIPS resubmission today and tomorrow, and don't think I will have a big enough consecutive timeslot to review before Thursday. Just so you know

robinholzi commented:
> hey Robin, thanks for the PR! I will go through it _probably_ on Thursday/Friday. I have to prepare teaching and a NeurIPS resubmission today and tomorrow, and don't think I will have a big enough consecutive timeslot to review before Thursday. Just so you know

Thanks Maxi! Thanks totally fine for me :) Let me know if you have any questions, I can always provide you some insights in person :)

MaxiBoether commented on modyn/supervisor/internal/pipeline_executor/pipeline_executor.py:
```diff
@@ -1,554 +1,855 @@
+
+        for replay_data, request_time in replay_data_generator:
+            # setting sample here to have correct logs in process_new_data
+            s.current_sample_time = min(
```
also, is this really necessary or not? Because the minimum search might be unnecessary compute for large batches

robinholzi replied:
I think it doesn't make a big difference what sample time we choose for a batch, it should just be one of the times in the batch. Let's use the time of the first sample, wdyt?

MaxiBoether replied:
I am still a bit confused. According to the docstring, it's "The unix timestamp of the last sample seen and processed by the pipeline executor". Shouldn't it be the maximum then? The role of this field is not clear to me yet. Is this just for logging/plotting purposes?

robinholzi replied:
Yes, it's only for logging/plotting. It should take the values of a representative sample in the current (training) batch. That allows to sort the pipeline stage execution logs by the dataset time dimension.

MaxiBoether commented on modyn/supervisor/internal/pipeline_executor/pipeline_executor.py:
```diff
@@ -1,554 +1,855 @@
-        self.trigger.inform_previous_model(model_id)
-        self.pipeline_log["supervisor"]["triggers"][trigger_id]["store_trained_model_time"] = self._sw.stop()
+    @pipeline_stage(PipelineType.NEW_BATCH, PipelineStage.EVALUATE_TRIGGER_POLICIES, track=True)
+    def _evaluate_trigger_policies(
```
nit: is there any reason this is plural? conceptually, there is just a single policy

robinholzi replied:
It would be possible to evaluate multiple policies before deciding. However, the final decision could then be again seen as a single, higher-order (aggregation) policy. So singular makes sense I guess.

MaxiBoether commented on modyn/supervisor/internal/pipeline_executor/pipeline_executor.py:
```diff
@@ -1,554 +1,855 @@
+        trigger_index = -1
+        trigger_indexes: list[int] = []
+        for i, trigger_index in enumerate(lazy_trigger_indexes):
+            # Run training and evaluation subpipelines
```
needs to be updated without subpipeline concept

MaxiBoether commented on modyn/supervisor/internal/pipeline_executor/pipeline_executor.py:
```diff
@@ -1,554 +1,855 @@
-            first_timestamp = triggering_data[0][1]
-            last_timestamp = triggering_data[-1][1]
+    @pipeline_stage(PipelineType.NEW_BATCH, PipelineStage.EXECUTE_TRIGGERS, track=True)
+    def _execute_triggers(
```
I am not sure whether I like the wording `execute` here. Maybe `handle` (because we _execute_ training jobs, while handling triggers), although handle is also quite general...

robinholzi replied:
maybe `invoke`?

robinholzi replied:
Okay no, let's just stick with `handle`

MaxiBoether commented on modyn/tests/supervisor/internal/pipeline_executor/test_pipeline_executor.py:
```diff
@@ -116,199 +139,222 @@ def teardown():
-            assert handle_mock.call_count == 6
-            assert get_new_data_mock.call_count == 3
+            assert handle_mock.call_count == 3
+            assert get_new_data_mock.call_count == 3 + 1
```
why the +1?

MaxiBoether commented on modyn/tests/supervisor/internal/pipeline_executor/test_pipeline_executor.py:
```diff
@@ -116,199 +139,222 @@ def teardown():
             assert handle_mock.call_args_list == expected_handle_mock_arg_list
 
-            expected_get_new_data_arg_list = [call("test", 21), call("test", 43), call("test", 45)]
+            expected_get_new_data_arg_list = [call("test", 21), call("test", 43), call("test", 45), call("test", 45)]
```
why the addition here?

MaxiBoether reviewed (changes requested):
Thank you Robin! My comments are mostly nitpicky. The biggest thing that needs to change is as discussed today to port the removal of the "multiple pipeline concept" here, and instead have this parent relationship. After addressing the open comments (esp in larger PRs, I prefer to handle reviews such that the author acknowledges everything that has been done with a comment like "addressed" or "done", and then the reviewer solves it when he's happy because otherwise stuff just gets lost) and porting this, I will do another round and then we can merge this :) thank you!

robinholzi commented on modyn/utils/timer.py:
```diff
@@ -0,0 +1,20 @@
+X = TypeVar("X")
+
+
+def timed_generator(generator: Generator[X, None, None]) -> Generator[tuple[X, float], None, None]:
```
These TypeVars behave a bit like C++ templates (without the whole compile-time concept though, they only serve type annotation purposes and don't change the runtime behavior).

The same concepts applies to decorators. these type annotations allow us that the returned result of the function/generator has exactly the same types in place of the TypeVars as the input object. With `typing.any` we would lose this information. This is annoying if you e.g. pass a complex data structure as `X` as you wouldn't get any IDE type hints for the returned result.

MaxiBoether replied:
Ahhh, I see. I thought the IDE would be able to infer the type with typing.Any (so i thought I would get exactly the behavior you mention using typing.Any. But then ok thanks for letting me know

robinholzi commented on modyn/tests/utils/test_timer.py:
```diff
@@ -0,0 +1,33 @@
+from modyn.utils.timer import timed_generator
+
+
+def test_timed_generator() -> None:
```
Yeah, it failed once or twice while my local system was under full load, in the CI it should be fine though as no other heavy processes are executed at the runners.

robinholzi commented on modyn/supervisor/internal/supervisor.py:
```diff
@@ -241,8 +241,9 @@ def start_pipeline(
-            return pipeline_res_msg(pipeline_id=pipeline_id, exception="Failed to execute pipeline")
+        except Exception as ex:  # pylint: disable=broad-except
+            raise ex
+            # return pipeline_res_msg(pipeline_id=pipeline_id, exception="Failed to execute pipeline")
```
Just a debugging artefact I forgot to remove, thanks for noticing!

robinholzi commented on modyn/supervisor/internal/pipeline_executor/models.py:
```diff
@@ -0,0 +1,406 @@
+    # stage specific log info
+    info: StageInfo | None = Field(None)
+
+    def online_df(self, extended: bool = False) -> pd.DataFrame | None:
```
The naming of `SelectorInformLog` is off, it should be `SelectorInformInfo`. Like `TrainingInfo`, `StoreEvaluationInfo`, ... it is additional logging info that only occurs at one certain stage and is therefore included in the `info` property of `StageLog` that has type `StageInfo`. `TrainingInfo`, `StoreEvaluationInfo`, etc. therefore all subclass `StageInfo` (directly or transitively).

The `stage_runs_df` property function of `SupervisorLogs` was indeed not used anymore and even relied on non-existent functions, it surprises me that the linting didn't catch that. I removed it completely.

robinholzi commented on modyn/supervisor/internal/pipeline_executor/pipeline_executor.py:
```diff
@@ -1,554 +1,855 @@
+from modyn.utils.timer import timed_generator
+from typing_extensions import Concatenate, ParamSpec
+
+from .models import (
```
I'll convert it to an absolute one. From my experience, it is desirable to use relative imports whenever possible (when that doesn't require going back beyond the current package, i.e. the parent). The benefit of that is that you see dependencies more clearly. Isort would e.g. also bundle them together in the last import block so "local" dependencies are visible more clearly.

robinholzi commented:
@MaxiBoether I think I've addressed all your comments, thanks for the suggestions. Once the CI is green you can checkout the diff in #439 (includes a couple of other tweaks that make the new logging format usable).

MaxiBoether commented on modyn/supervisor/internal/pipeline_executor/models.py:
```diff
@@ -0,0 +1,499 @@
+
+    # incremental logs
+
+    supervisor: SupervisorLogs = Field(default_factory=SupervisorLogs)
```
rename to supervisor_logs? was thinking for a moment that this owns a superviosr object now.

robinholzi replied:
Will do! I thought that would be clear from the typing, but fair point.

MaxiBoether commented on modyn/supervisor/internal/evaluation_result_writer/json_result_writer.py:
```diff
@@ -17,7 +18,19 @@ def add_evaluation_data(self, dataset_id: str, dataset_size: int, evaluation_dat
 
-    def store_results(self) -> None:
+
+class JsonResultWriter(_JsonResultWriter):
```
I don't get the purpose of this yet

robinholzi replied:
Previously we wrote all evaluation calls to dedicated files. The default should now be not to create one file per evaluation request but use the central json that is built by the `pipeline_executor`. To keep the old functionality i renamed the old JsonResultWriter to `DedicatedJsonResultWriter` while the new one doesn't create files on its own.

MaxiBoether reviewed (approved):
Generally, to the best I can overlook this, this looks good. Some minor things are still open.

I am mostly confused about the dfs. We have `df()` on the `SupervisorLogs`, we have the `online_df()` function where I am a little confused by the name, and the relationship of the dfs and logs is not 100% clear yet. Maybe it would be helpful if you could write a paragraph somewhere on how logging works now? Mainly on the dfs, the pipeline log and what their relationship is, when what gets materialized etc. I think that would be helpful. Thank you, we're getting close!

robinholzi commented on modyn/supervisor/internal/pipeline_executor/models.py:
```diff
@@ -0,0 +1,499 @@
+
+    def online_df(self) -> pd.DataFrame | None:
+        """
+        Some stages might want to store data in the online DataFrame.
```
I will strip the online from the name and add a global documentation `docs/LOGGING.md` that explains how logging works and how one can use it.

robinholzi commented on modyn/supervisor/internal/pipeline_executor/models.py:
```diff
@@ -0,0 +1,499 @@
+# pylint: disable=E1133
```
Not sure anymore, I think we can drop it.

robinholzi commented:
@XianzheMa I'm happy to give you a short intro to the new logging setup in case the `LOGGING.md` documentation isn't sufficient :)

robinholzi commented on modyn/supervisor/internal/grpc_handler.py:
```diff
@@ -276,69 +278,47 @@ def start_training(
             if (
-                "interval" not in pipeline_config["training"]["checkpointing"]
-                or "path" not in pipeline_config["training"]["checkpointing"]
+                pipeline_config["training"]["checkpointing"].get("interval") is None
```
Probably I'm not always consistent, sometimes I just want to make it more explicit (e.g. when using the value for more than just `if` blocks).

robinholzi commented:
@MaxiBoether Huge thanks for the effort reviewing this 💯
